### PR TITLE
Add Extract support to client-go apply builders

### PR DESCRIPTION
**Pull request by jpbetz**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

This PR is the last step toward implementing [Apply for client-go's typed client](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2155-clientgo-apply)

umbrella issue: https://github.com/kubernetes/enhancements/issues/2155

**Special notes for your reviewer**:

A full schema is required to "extract" an apply configuration from an object. The schema contains information about which fields are atomic and granular, what fields are map keys. Without the schema it is not possible to correctly combine the managed field data with the object to create an apply configuration. We've decided to embed the schema in the generated client as a string. The alternatives were: (1) embed it as a string, (2) download/cache it from the api server or (3) move the extract logic server side.  Alternative 1 was the only one that seemed tractable in the short term. Performance impact: takes about 45ms to load the schema for the first time (we lazy load it).

**How this PR avoids dependency cycle pitfalls**: Instead of depending on `swagger.json`, this depends on a build artifact that is produced earlier in the build chain-- the openAPI `definitions` created by `openapi-gen` and written to `zz_generated.openapi.go`. This artifact includes only the `definitions` section of an OpenAPI spec, but not the `endpoints` (`swagger.json` has both). Since the definitions are all that is needed for this PR, it works just fine. Moving `update-openapi-spec` to run before `update-codegen` was not an option. That would have introduced a cycle because `update-openapi-spec` runs the kube-apiserver and curls it to create `swagger.json`, and the kube-apiserver build process depends on the code `update-codegen` produces.

The SMD version bump picks up https://github.com/kubernetes-sigs/structured-merge-diff/pull/185

Generated code is in a separate commit.

**Does this PR introduce a user-facing change?**:

Release note will be added when all the last change in the umbrella issue is merged

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

- [KEP]: [KEP 2155](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2155-clientgo-apply)

```docs
TODO
```

/sig api-machinery
/wg api-expression

### CODE REVIEW

**Comment by jpbetz** [comment]
cc @apelisse @kevindelgado

**Review by kevindelgado — COMMENTED** [review]
Looks good Joe! One little optional suggestion.

File: staging/src/k8s.io/code-generator/cmd/applyconfiguration-gen/generators/applyconfiguration.go
Comment (jpbetz) [reply staging/src/k8s.io/code-generator/cmd/applyconfiguration-gen/generators/applyconfiguration.go:326]: Yes, good catch. I'll add it.

File: staging/src/k8s.io/code-generator/cmd/applyconfiguration-gen/generators/applyconfiguration.go
Comment (Jefftree) [inline staging/src/k8s.io/code-generator/cmd/applyconfiguration-gen/generators/applyconfiguration.go:323]: Out of curiosity, why nil instead of an empty applyconfig? Calling this functions probably means the caller has the intent of creating an applyconfig regardless of the managedfields. Wouldn't it be easier to just create it here?

Comment (jpbetz) [reply staging/src/k8s.io/code-generator/cmd/applyconfiguration-gen/generators/applyconfiguration.go:323]: I see your point. I was thinking more about giving a signal to users to let them know that no managed fields were found for the field manager they provided.
 
Unfortunately, it not possible to distinguish between (1) didn't find any managed fields because the given fieldManager has never applied to the object and (2) didn't find any managed fields because all the fields owned by the field manager have been claimed by other managers.

If the user did manage field before, it's more likely they want to create the object regardless of what's happened to their field ownership since they last applied.

I suppose that if we don't provide a signal to indicate if the managed fields were found or not, they could always check the object directly to get the answer.

File: test/integration/client/client_test.go
Comment (jpbetz) [reply test/integration/client/client_test.go:947]: thx, yes, this is incorrect. I'll just deref so they can't get out of sync

File: staging/src/k8s.io/code-generator/cmd/applyconfiguration-gen/generators/applyconfiguration.go
Comment (jpbetz) [reply staging/src/k8s.io/code-generator/cmd/applyconfiguration-gen/generators/applyconfiguration.go:323]: I've updated to run an empty object and clarified exactly what is returned and under what conditions it can be returned (because it didn't seem 100% obvious to me that developers would know this).

**Comment by jpbetz** [comment]
Thanks for the reviews @Jefftree @kevindelgado! Feedback applied. PTAL.

File: staging/src/k8s.io/apimachinery/pkg/util/managedfields/extract.go
Comment (kevindelgado) [inline staging/src/k8s.io/apimachinery/pkg/util/managedfields/extract.go:34]: *It is possible

File: staging/src/k8s.io/code-generator/cmd/applyconfiguration-gen/generators/applyconfiguration.go
Comment (kevindelgado) [inline staging/src/k8s.io/code-generator/cmd/applyconfiguration-gen/generators/applyconfiguration.go:325]: It is

File: staging/src/k8s.io/code-generator/cmd/applyconfiguration-gen/generators/internal.go
Comment (kevindelgado) [inline staging/src/k8s.io/code-generator/cmd/applyconfiguration-gen/generators/internal.go:31]: *internalGenerator?

**Comment by deads2k** [comment]
> The alternatives were: (1) embed it as a string, (2) download/cache it from the api server or (3) move the extract logic server side. #1 was the only option that seemed tractable in the short term.

This seems reasonable for generated clients.  I'd like to see a way to use this with a dynamic client in the next release.  I suspect that will mean finding a way to get the required schema from the openapi doc.

File: test/integration/client/client_test.go
Comment (soltysh) [inline test/integration/client/client_test.go:986]: As mentioned in the slack, these tests should cover at least:
1. changing single element, replicas, for example, my example from https://gist.github.com/soltysh/7f5e924e6dee5e69a1ab00c6a72012b1#file-test3-go-L82 fails with validation error b/c labels are removed.
2. updating status (this is what majority of controllers does).

**Review by soltysh — CHANGES_REQUESTED** [review]

File: test/integration/client/client_test.go
Comment (jpbetz) [reply test/integration/client/client_test.go:1166]: Thank

> > The alternatives were: (1) embed it as a string, (2) download/cache it from the api server or (3) move the extract logic server side. #1 was the only option that seemed tractable in the short term.
> 
> This seems reasonable for generated clients. I'd like to see a way to use this with a dynamic client in the next release. I suspect that will mean finding a way to get the required schema from the openapi doc.

Yeah, @lavalamp had mentioned needing to support the dynamic client as well. We'll follow up.

File: hack/update-codegen.sh
Comment (jpbetz) [reply hack/update-codegen.sh]: Thanks for catching this. I spent quite a while reviewing how everything works and I've come up with a solution that seems to avoid falling into any dependency cycle pits:

Instead of depending on `swagger.json`, I'm going to depend on a build artifact that is produced much earlier-- the openAPI `definitions` that are produced by `openapi-gen` and written to `zz_generated.openapi.go`. This artifact includes only the `definitions` (aka models, aka schemas) and not the `endpoints`, but that's all we need.

This avoids introducing a cycle between `update-openapi-spec` and `update-codegen`. Moving `update-openapi-spec` to run before `update-codegen` would have introduced a cycle because `update-openapi-spec` runs the kube-apiserver and curls it to create `swagger.json`, and the kube-apiserver build process depends on the code `update-codegen` produces.

**Comment by jpbetz** [comment]
> > The alternatives were: (1) embed it as a string, (2) download/cache it from the api server or (3) move the extract logic server side. #1 was the only option that seemed tractable in the short term.
> 
> This seems reasonable for generated clients. I'd like to see a way to use this with a dynamic client in the next release. I suspect that will mean finding a way to get the required schema from the openapi doc.

Yeah, @lavalamp had mentioned needing to support the dynamic client as well. We'll follow up.

File: test/integration/client/client_test.go
Comment (jpbetz) [reply test/integration/client/client_test.go:1166]: I've added a bunch more test cases to both cover these cases and to improve how well the tests document the functionality.

File: pkg/generated/openapi/cmd/models-schema/main.go
Comment (deads2k) [inline pkg/generated/openapi/cmd/models-schema/main.go:30]: really?  I thought we already had this.

Comment (apelisse) [reply pkg/generated/openapi/cmd/models-schema/main.go:30]: Today, the best way to generate the openapi is to run an apiserver and get `/openapi/v2` (which is what we do in `hack/update-openapi-spec.sh`).

Comment (jpbetz) [reply pkg/generated/openapi/cmd/models-schema/main.go:30]: `swagger.json` is a downstream artifact of this that combines this information with endpoint information to construct the complete OpenAPI.  This artifact only has information about the `definitions` (the types, not the endpoints). The problem is that depending directly on `swagger.json` introduces a dependency cycle. I put a section in the description summarizing.

**Comment by jpbetz** [comment]
@liggitt could I get your approval for the SMD version bump?

**Comment by liggitt** [comment]
/approve
for dep bump, but needs rebase

**Comment by apelisse** [comment]
@liggitt Do you to specifically merge that last, after everything else is merged so that it doesn't disrupt other people?

If I'm correct, the disruption is that we need to re-run a script to update some generated code.
If all existing API changes (let's say there are n) can go through without being disrupted, then we would merge that last and only run the script once. If we merge that first, everyone would have to re-run the script (n times), so that would seem better.

But in practice, don't people have to re-run a bunch of scripts anytime an API change happens anyway? So technically, people will have to run the script n^2 times.

**Comment by jpbetz** [comment]
> note this will disrupt PRs in flight with [API changes](https://github.com/kubernetes/kubernetes/pulls?q=is%3Apr+is%3Aopen+label%3Akind%2Fapi-change+milestone%3Av1.21+label%3Aapi-review+) ... dropping something this disruptive on code freeze day is less than ideal

I'm okay with holding this and doing a rebase after any in-flight API change PRs merge.

**Comment by jpbetz** [comment]
/hold
Until all [API changes](https://github.com/kubernetes/kubernetes/pulls?q=is%3Apr+is%3Aopen+label%3Akind%2Fapi-change+milestone%3Av1.21+label%3Aapi-review+) merge (see https://github.com/kubernetes/kubernetes/pull/99759#issuecomment-794087562)

**Comment by liggitt** [comment]
> Until all [API changes](https://github.com/kubernetes/kubernetes/pulls?q=is%3Apr+is%3Aopen+label%3Akind%2Fapi-change+milestone%3Av1.21+label%3Aapi-review+) merge (see [#99759 (comment)](https://github.com/kubernetes/kubernetes/pull/99759#issuecomment-794087562))

yeah, at least until the [ones with lgtm+approve](https://github.com/kubernetes/kubernetes/pulls?q=is%3Apr+is%3Aopen+label%3Akind%2Fapi-change+milestone%3Av1.21+label%3Algtm+label%3Aapproved+) currently in the merge pool are in... the others likely have outstanding comments to address and will need another push anyway

**Comment by jpbetz** [comment]
Rebased

**Comment by jpbetz** [comment]
/hold cancel
None of the in-flight [approved+lgtmd API changes]( [API changes](https://github.com/kubernetes/kubernetes/pulls?q=is%3Apr+is%3Aopen+label%3Akind%2Fapi-change+milestone%3Av1.21+label%3Aapi-review+)) have any schema changes. This looks like a good time to merge this

**Comment by jpbetz** [comment]
Rebased to resolve modules.txt conflicts

Would someone please retag?

**Comment by jpbetz** [comment]
Had to rebase again

**Comment by liggitt** [comment]
/hold

This is failing batch merges, looks like it has a non-conflicting merge error with master

**Comment by jpbetz** [comment]
Having a look

**Comment by liggitt** [comment]
flake was #99979

**Comment by BenTheElder** [comment]
https://github.com/kubernetes/kubernetes/issues/98854
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/99759/pull-kubernetes-e2e-kind/1369859724211654656
and maybe a new one:
```

Kubernetes e2e suite: [sig-network] EndpointSlice should create Endpoints and EndpointSlices for Pods matching a Service [Conformance] expand_less | 3m1s
-- | --
/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:630 Mar 11 04:26:32.486: timed out waiting for Pods to have IPs assigned Unexpected error:     <*errors.errorString \| 0xc00024a250>: {         s: "timed out waiting for the condition",     }     timed out waiting for the condition occurred /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/endpointslice.go:317

```
/retest

**Comment by BenTheElder** [comment]
https://github.com/kubernetes/kubernetes/issues/98854 again
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/99759/pull-kubernetes-e2e-kind/1369886294947139584
/retest

**Pull request merged**
